accepted/00000-salt-ssh-with-salt-bundle.md

```diff
@@ -0,0 +1,95 @@
+
+Salt Bundle is a single package containing salt codebase, Python and python modules required for salt.
+
+# Motivation
+[motivation]: #motivation
+
+Due to the deprecation of Python 2 and dropping support of it, while some of supported Linux systems still has no proper Python 3 version supported by Salt, we need to have a way to process Salt-SSH events on the minion side.
+
+The following systems are using salt codebase older than `3002.2`:
+- SLE 11 (LTSS ends 31 Mar 2022, no plans to support it with Salt Bundle)
+- SLE 12
+- RHEL 7 (and clones)
+- Debian 9
+
+We can deploy Salt Bundle with bundled Python inside the virtual environment and handle all the Salt-SSH events using Salt Bundle with making Salt-SSH independent from Python on the minion.
+Additionally we are extending the list of working modules as Salt Bundle contains binary python modules required by some of Salt modules.
+
+This approach will help us to limit the number of supported versions of Salt codebase and drop `py26-compat-salt`, `py27-compat-salt` (as these packages are shipping `2016.11.10` and `3000.3` respectively) and all Python 2 dependencies maintenance.
+So we don't need to maintain old salt codebases, but focus only on the latest one.
+
+# Detailed design
+[design]: #detailed-design
+
+## Uyuni roster
+
+We can use Uyuni roster module to provide roster data instead of creating flat roster files with Java. PR moving the logic to Uyuni roster on Java side: https://github.com/uyuni-project/uyuni/pull/4457
+
+This change can help to avoid creating temporary roster files (https://docs.saltproject.io/en/latest/topics/ssh/roster.html) and implements caching additionally.
+
+## Pre flight script
+
+Salt can execute a shell script on the salt-ssh client before the "real salt-ssh execution" starts.
+This script downloads the salt-bundle package from the appropriate bootstrap repository and extracts the virtual environment.
+By-default there is no way to pass additional parameters to pre fligt script running on the minion.
+In this case we have to generate individual pre flight script to each of the minion. The main drawback here is maintaining these pre flight script files.
+
+The other solution is to pass additional parameter to the static pre flight script.
+The pre flight script calculates the bootstrap repository URL in a similar manner to the existing boostrap.sh script, based on the operating system and OS architecture of the system it is executed on.
+It requires one parameter from the Uyuni server: host, port (entry point) for different contact methods and pathes (SSH from server/proxy, direct access to the repos or with the tunnel).
+The salt codebase need to be modified to pass additional parameters, therwise we have to generate pre flight script individually for each minion.
+
+## Deploying Salt Bundle to the minion
+
+We don't need to install Salt Bundle package on the minion.
+We need to extract the virtual environment of the bundle from the package.
```

Comment:
What are the pros and cons of actually installing the package?

Reply:
For some of the usecases users may use the minion connected to the server with venv-salt-minion, but use it with the other system as salt-ssh minion with handling salt-ssh with salt bundle. I don't think we should really support such cases. The main reason is that we are not installing any package on salt-ssh minions, so why to change this with salt bundle with salt-ssh approach? Most of the users expects salt-ssh to be agentless management method, in case of installing the package the agent will be installed, but not used. And additionally: salt bundle package also deploys pkgset beacon notification plugins, configs etc. which are not required for salt-ssh.